### ISSUE TITLE
RegisterHttpRequestMessage fails with Azure App Service Mobile App

### ISSUE THREAD

**Issue by tillig** [autofac/Autofac.WebApi#3]
_From @Louliloul on June 22, 2015 16:8_

Hi !

I'm currently playing with Azure App Services Mobile App. I set up a basic solution in VS 2015 RC with Autofac as DI, and one of my dependencies requires the current HttpRequestMessage. So i used the extensions RegisterHttpRequestMessage in order to be able to inject it as needed. But the call to RegisterHttpRequestMessage produces the following exception:

Exception Details: System.TypeAccessException: Attempt by security transparent method 'Autofac.Integration.WebApi.RegistrationExtensions.RegisterHttpRequestMessage(Autofac.ContainerBuilder, System.Web.Http.HttpConfiguration)' to access security critical type 'System.Web.Http.HttpConfiguration' failed.

Assembly 'Autofac.Integration.WebApi, Version=3.0.0.0, Culture=neutral, PublicKeyToken=17863af14b0044da' is marked with the AllowPartiallyTrustedCallersAttribute, and uses the level 2 security transparency model.  Level 2 transparency causes all methods in AllowPartiallyTrustedCallers assemblies to become security transparent by default, which may be the cause of this exception.

As the call to RegisterHttpRequestMessage comes before the call to the concerned module, my controller instanciation fails. The second breakpoint below is never reached.
![image](https://cloud.githubusercontent.com/assets/7660709/8286395/7b47021e-1909-11e5-8219-3ce66e14dbad.png)

And when i try to access the service homepage, it tells that the service is unhealthy.
![image](https://cloud.githubusercontent.com/assets/7660709/8286191/70ada066-1908-11e5-9e2a-1a84be964123.png)

_Copied from original issue: autofac/Autofac#649_

**Comment by tillig** [comment]
Three questions:

1. What framework are you targeting? .NET 4.5? DNX? .NET 4.6?
2. I see the Autofac.Integration.WebApi assembly version number is `3.0.0.0` in the error. `3.1.0.0` came out in July 2013. Have you tried updating? (Not promising that will fix it, just curious if running the latest bits will help.)
3. What other packages do you have installed? Can we see your `packages.config`?

**Comment by Louliloul** [comment]
Thanks for your reply.

1. .NET framework 4.5
2. Actually the Autofac.Integration.WebApi is labeled 3.1.0 in packages.config, and it's the latest available in NuGet manager The solution is rather new, I created it 1 week ago.
3.packages.config
![image](https://cloud.githubusercontent.com/assets/7660709/8289842/332a9b7e-191f-11e5-9886-79fa1ffeb439.png)
![image](https://cloud.githubusercontent.com/assets/7660709/8289862/58c489a8-191f-11e5-88aa-8efa0f7a3ba6.png)

**Comment by Louliloul** [comment]
In my project, in the references, i just realized that the version of Autofac.Integration.WebApi in the properties panel is 3.0.0.0.
In the nuget package that was installed, in the lib folder, there is only one folder "net40"

**Comment by tillig** [comment]
Sorry it's taken so long to get back to you. I don't generally work on Azure apps so I had to set up an environment. I think I figured out the issue while I was working through [the demo mobile services app](https://azure.microsoft.com/en-us/documentation/articles/mobile-services-dotnet-backend-windows-store-dotnet-get-started/).

**Try using the `Autofac.WebApi2` package instead of `Autofac.WebApi`.** There were some breaking changes when MS revised Web API so we had to issue a corresponding package, the latest version of which (right now) is 3.4.0. I was able to successfully register the `HttpRequestMessage` in the demo app without issue using that package.

I'm going to close this since I'm pretty sure that's the answer. If it is still not working, let me know and maybe post a repro so I can see it in action and we can re-open.
